### ISSUE TITLE
[Feature Request] Allow for QTPIE charge equilibration with external efield

### ISSUE THREAD

**Issue by navlalli** [lammps/lammps#4226]
**Summary**

LAMMPS currently supports `fix qeq/reaxff` and `fix acks2/reaxff`, which allow for charge equilibration with an external electric field. It could be valuable to add `fix qtpie/reaxff` to the REAXFF package to allow for charge equilibration using the [QTPIE](https://doi.org/10.1016/j.cplett.2007.02.065) charge equilibration scheme. 

**Detailed Description**

When using [QEq](https://pubs.acs.org/doi/10.1021/j100161a070) charge equilibration with pair_style reaxff and when applying an external electric field, the dynamics of the simulation become unrealistic due to long range charge transfers predicted by the QEq method. For example, molecules that should be neutral acquire a net charge and are driven to translate by the electric field. This has been discussed by [Kritikos et al](https://doi.org/10.1016/j.combustflame.2022.112238) and [Koski et al](https://pubs.acs.org/doi/10.1021/acs.jctc.1c00975). QTPIE charge equilibration can successfully shield charge transfers (see, e.g., [Kritikos et al](https://pubs.acs.org/doi/full/10.1021/acs.jpca.0c08040)) and has a similar formulation to QEq, which means that it can be, to a reasonable approximation, used with the same REAXFF that QEq has been parameterised for.  Furthermore, QTPIE can be formulated to allow for periodic boundaries in the direction of an external electric field.

I would like to try and implement this myself. I saw that work has been completed on this before by @stanmoore1 at https://github.com/stanmoore1/lammps/tree/qtpie . I wanted to know if there were any particular obstacles that prevented the implementation of QTPIE from being completed or anything in particular that I should be aware of before giving it a go?

Thanks!

**Comment by stanmoore1** [comment]
@navlalli See the discussion here: https://matsci.org/t/qtpie-how-to-use-qtpie-in-lammps/48192. @srtee may be able to help as mentioned in that thread. Let me know if you have specific questions.

**Comment by navlalli** [comment]
@stanmoore1 Thank you for the link. I have a question on the units of `chi_field`. According to [Koski et al](https://pubs.acs.org/doi/10.1021/acs.jctc.1c00975), an external electric field is introduced into QEq by replacing $\chi_i$ with $`\chi_{i} - \vec{r_i} \cdot \vec{E}_{\mathrm{el}}`$, where $\chi_i$ is the Mulliken electronegativity of atom $i$, $\vec{r_{i}}$ is the position vector of atom $i$, and $`\vec{E}_{\mathrm{el}}`$ is the external electric field. Since $`\chi_{i}`$ is in units of eV (see the [docs](https://docs.lammps.org/fix_qeq_reaxff.html)), wouldn't $`\vec{r_i} \cdot \vec{E}_{\mathrm{el}}`$ also need to be in the same units somehow?

When looking at  https://github.com/lammps/lammps/blob/aa0b6c47c20cca07a7b2c63479ed7dd998094d3b/src/REAXFF/fix_qeq_reaxff.cpp#L1128-L1129
https://github.com/lammps/lammps/blob/aa0b6c47c20cca07a7b2c63479ed7dd998094d3b/src/fix_efield.cpp#L66
https://github.com/lammps/lammps/blob/aa0b6c47c20cca07a7b2c63479ed7dd998094d3b/src/REAXFF/fix_qeq_reaxff.cpp#L1150
it seems like `chi_field`, which is $- \vec{r_i} \cdot \vec{E}_{\mathrm{el}}$, is in units of volts.

Perhaps I am overlooking something, but I would appreciate an explanation of how this works?

**Comment by navlalli** [comment]
I now understand where I was going wrong. It's because the charge in real units is a multiple of electron charge, so $`\vec{r_i} \cdot \vec{E}_{\mathrm{el}}`$ must be multiplied by the electron charge to account for this, which results in an energy in joules. Then this energy must be divided by electron charge to convert to eV. Therefore, $`\vec{r_i} \cdot \vec{E}_{\mathrm{el}}`$ is already the energy in eV.

**Comment by srtee** [comment]
Good job working it out. I feel like getting electric units consistent was one of the hardest parts of working on the CONP and ELECTRODE codes -- I wish there were a better way.

**Comment by navlalli** [comment]
@srtee I've got a quick question on the sign of chi_field.

So the sign on line 1150 makes sense to me as this is equivalent to $- \vec{r_i} \cdot \vec{E}_{\mathrm{el}}$

https://github.com/lammps/lammps/blob/aa0b6c47c20cca07a7b2c63479ed7dd998094d3b/src/REAXFF/fix_qeq_reaxff.cpp#L1129

https://github.com/lammps/lammps/blob/aa0b6c47c20cca07a7b2c63479ed7dd998094d3b/src/REAXFF/fix_qeq_reaxff.cpp#L1150

But then I'm confused about the sign of chi_field on line 1157 when using the atom-style potential. I thought the potential for each atom (i.e., `efield->efield[i][3]`) in a uniform field would also be $`- \vec{r_i} \cdot \vec{E}_{\mathrm{el}}`$ since  $`\vec{E}_{\mathrm{el}} = - \nabla \phi`$, where $\phi$ is the electric potential. Therefore, I am unsure why there is a negative sign on line 1157?

https://github.com/lammps/lammps/blob/aa0b6c47c20cca07a7b2c63479ed7dd998094d3b/src/REAXFF/fix_qeq_reaxff.cpp#L1157

**Comment by srtee** [comment]
You're probably right (that the minus sign shouldn't be there in the last line) -- I may have somehow written my tests / inputs wrongly and gotten that sign wrong. Best to write a test of some kind that we can use to avoid regression in future updates.

I am currently swamped with teaching but I can come back to verifying this two or three weeks from now.

**Comment by navlalli** [comment]
Thanks for the quick response. I can take a look at it next week! I'll do a quick test with a water molecule and see if the dipole moment changes as expected, i.e., the magnitude of the dipole moment should be larger when the sign is correct compared to when the sign is wrong.

**Comment by srtee** [comment]
A better test is a set of inputs imposing _identical_ uniform fields with all options (raw numbers, equal-style, per-atom, and field vs potential) and then asserting identical induced charge for each (or debugging any differences).

Equalities are more easily asserted or debugged than inequalities ;)

For context -- I made these code modifications contributing to a paper that I finally withdrew authorship from just before review approval (for personal reasons -- I was in a terrible headspace this time last year). It was used to study ReaxFF under high-frequency alternating fields, so unfortunately our results wouldn't have incisively revealed this sign error.

**Comment by navlalli** [comment]
That sounds like a better test! I'll get onto that next week.

I see, makes sense.

**Comment by navlalli** [comment]
I have done this test with a water molecule using an input script containing

```
if "${etype} == const" then &
    "fix efield all efield 0.0 0.0 0.1" &
elif "${etype} == equal" &
    "variable ef equal 0.1" &
    "fix efield all efield 0.0 0.0 v_ef" &
elif "${etype} == atom" &
    "variable afor atom 0.1" &
    "variable apot atom -0.1*z" &
    "fix efield all efield 0.0 0.0 v_afor potential v_apot" &
else &
    "print 'Wrong etype'" &
    quit
```

and have found that the charges are the same for the three methods of imposing an electric field using `fix efield` when changing the sign of line 1157. I have therefore included this sign change in pull request #4312 . Ideally, I should add this as a unit test into lammps, but I won't be able to this week due to time constraints.

I have attached the files in case you are interested.
[test.zip](https://github.com/user-attachments/files/17189294/test.zip)
